src/v2/view-builder/internals/BaseForm.js

```diff
@@ -111,39 +111,45 @@ export default Form.extend({
       }
     }
   },
-
+  
   /*
-   * Whether an form error message should be rendered as the default BaseForm error banner or not
-   * if this function returns true, BaseForm won't render the error banner after form submit. Instead,
-   * the form needs to override showCustomErrorCallout with the logic to render that error message
-   */
-  isErrorCalloutCustomized() {
-    return false;
-  },
-
-  /*
-   * Renders an error message (if any) after form is submitted
-   * This should be overriden if isErrorCalloutCustomized() returns true
-   */
-  showCustomErrorCallout() {
-    this.$('.o-form-error-container').addClass('o-form-has-errors');
-  },
+  * Views should override this function to render custom error callouts for invalid form actions.
+  */
+  showCustomFormErrorCallout: _.identity,
 
   /*
-   * Renders the contents of messages object (if any) during initialize
+   * Renders the contents of messages object (if any, on error) during initialize
    * This function is called during Form.initialize, and will display
-   * messages when the form renders.
-   * Note: any errors happening after the form is rendered won't use this function,
-   * this is only used during initialize
+   * messages when the form reloads.
+   * Note: Any user action related form errors handled by FormController.showFormErrors
    */
-  showMessages() {
-    // render messages as text
-    const messagesObjs = this.options.appState.get('messages');
-    if (messagesObjs?.value.length) {
-      const content = messagesObjs.value.map((messagesObj) => {
-        return messagesObj.message;
+  showMessages(options) {
+    const messages = this.options.appState.get('messages') || {};
+    const errContainer = '.o-form-error-container';
+    if (Array.isArray(messages.value)) {
+      this.add('<div class="ion-messages-container"></div>', errContainer);
+      messages.value.forEach(obj => {
+        if(!obj?.class || obj.class === 'INFO') {
+          // add message as text
+          this.add(`<p>${obj.message}</p>`, '.ion-messages-container');
+        } else {
+          if(options instanceof View) {
+            // add custom callout or any view
+            this.add(options, errContainer);
```

Comment:
this means that for each of the messages items we will add the same `options` view? If the `options` argument is a View, maybe we don't need to render one per `messages` items? 

If I am following this right, let's say `options` is a `View`, and `messages` have 2 "ERROR" items, we will not render both errors, but instead we will render the `View` twice...

Reply:
The idea in widget is to show just one error if there are multiple error in the same IDX response.. Once the user resolve that, then show the next error if it still there.. According to UX we should not overwhelm users with many errors at a time..

But overtime, this `showMessages` got some hacky implementation in `TerminalView` ([here](https://github.com/okta/okta-signin-widget/blob/master/src/v2/view-builder/views/TerminalView.js#L118)) which adds more messages to `message` object.. to renders them as just plain text.. That is why i have to keep `forEach`..  But ideally it will always be the `top` error..


>but instead we will render the View twice...
There is a rare possibility.. But again there should be just one error message in the IDX api response. Let me know if you have example of IDX API response that send `two` error messages.

Reply:
>but instead we will render the View twice...

Actually i removed it that. Now `else if` will take care of add view just once.

Reply:
I don't have an example unfortunately on the top of my head, but if `messages` is an array, we should not assume it is always length=1, but instead we should iterate through all the items and display them... 
With that said I agree with not showing 2 callouts, that would look bad, but we should not ignore the second error if it comes. Perhaps we can show them as `bullets` of the same callout? Can be a refactor in a different PR if that makes more sense